python/pyiceberg/cli/console.py

```diff
@@ -0,0 +1,264 @@
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+import os
+from typing import Literal, Optional, Tuple
+
+import click
+from rich.console import Console
+from rich.table import Table
+from rich.tree import Tree
+
+from pyiceberg.catalog import Catalog
+from pyiceberg.catalog.hive import HiveCatalog
+from pyiceberg.catalog.rest import RestCatalog
+from pyiceberg.cli.output import ConsoleOutput, JsonOutput, Output
+from pyiceberg.exceptions import NoSuchNamespaceError, NoSuchTableError
+
+
+@click.group()
+@click.option("--catalog", default="catalog")
+@click.option("--output", type=click.Choice(["text", "json"]), default="text")
+@click.option("--uri", default=lambda: os.environ.get("PYICEBERG_URI"))
+@click.option("--credential", default=lambda: os.environ.get("PYICEBERG_CREDENTIAL"))
+@click.pass_context
+def run(ctx, catalog: str, output: str, uri: str, credential: str):
+    console = Console()
+    console.print("Welcome to PyIceberg 🐧")
+    ctx.ensure_object(dict)
+    if uri.startswith("http"):
+        ctx.obj["catalog"] = RestCatalog(name=catalog, properties={}, uri=uri, credential=credential)
+    elif uri.startswith("thrift"):
+        ctx.obj["catalog"] = HiveCatalog(name=catalog, properties={}, uri=uri)
+    else:
+        raise ValueError("Could not determine catalog type from uri. REST (http/https) and Hive (thrift) is supported")
+
+    if output == "text":
+        ctx.obj["output"] = ConsoleOutput()
+    elif output == "json":
+        ctx.obj["output"] = JsonOutput()
+    else:
+        raise ValueError(f"Unknown output: {output}")
+
+
+def _get_catalog_and_output(ctx) -> Tuple[Catalog, Output]:
+    """
+    Small helper to set the types
+    """
+    return ctx.obj["catalog"], ctx.obj["output"]
+
+
+@run.command()
+@click.pass_context
+@click.argument("parent", required=False)
+def list(ctx, parent: Optional[str]):
+    catalog, output = _get_catalog_and_output(ctx)
+
+    # still wip, will become more beautiful
+    # https://github.com/apache/iceberg/pull/5467/
+    # has been merged
+    if parent:
+        identifiers = catalog.list_tables(parent)
+    else:
+        identifiers = catalog.list_namespaces()
+
+    output.identifiers(identifiers)
+
+
+@run.command()
+@click.argument("identifier")
+@click.pass_context
+def describe(ctx, identifier: str):
+    """Describes a table"""
+    catalog, output = _get_catalog_and_output(ctx)
+
+    try:
+        table = catalog.load_table(identifier)
+        output.describe_table(table)
+    except NoSuchNamespaceError as exc:
+        output.exception(exc)
+
+
+@run.command()
+@click.option("--type", type=click.Choice(["all", "namespace", "table"]), default="all")
```

Comment:
Should it be `any` rather than `all`? There should be only one resource so we won't ever have 2 things to show.

Reply:
Good point, I've changed it to `--all`. We can also change it to `--table` and `--namespace`, but I like how this shows up as an option:
```
➜  python git:(fd-python-cli) ✗ pyiceberg properties --help
Usage: pyiceberg properties [OPTIONS] IDENTIFIER

  Fetches the properties from the namespace/table

Options:
  --type [any|namespace|table]
  --help                        Show this message and exit.
```